lib/portage/tests/util/file_copy/test_copyfile.py

```diff
@@ -42,25 +42,27 @@ def testCopyFileSparse(self):
             # files too big, in case the filesystem doesn't support
             # sparse files.
             with open(src_path, "wb") as f:
+                f.seek(2**16, os.SEEK_SET)
                 f.write(content)
-                f.seek(2**17, 1)
-                f.write(content)
-                f.seek(2**18, 1)
+                f.seek(2**17, os.SEEK_SET)
                 f.write(content)
                 # Test that sparse blocks are handled correctly at
-                # the end of the file (involves seek and truncate).
-                f.seek(2**17, 1)
+                # the end of the file.
+                f.truncate(2**18)
 
             copyfile(src_path, dest_path)
 
             self.assertEqual(perform_md5(src_path), perform_md5(dest_path))
 
-            # This last part of the test is expected to fail when sparse
-            # copy is not implemented, so mark it xfail:
-            pytest.xfail(reason="sparse copy is not implemented")
+            src_stat = os.stat(src_path)
+            dest_stat = os.stat(dest_path)
+
+            self.assertEqual(src_stat.st_size, dest_stat.st_size)
 
             # If sparse blocks were preserved, then both files should
             # consume the same number of blocks.
-            self.assertEqual(os.stat(src_path).st_blocks, os.stat(dest_path).st_blocks)
+            # This is expected to fail when sparse copy is not implemented.
+            if src_stat.st_blocks != dest_stat.st_blocks:
+                pytest.xfail(reason="sparse copy is not implemented")
```

Comment:
We could possibly do things like use an `fcntl.ioctl` wrapper to check for a successful FICLONE ioctl call, and then adjust assertions accordingly, like can we assert that FICLONE results in a sparse copy?

Reply:
I was thinking of something roughly like this:
```diff
diff --git a/lib/portage/tests/util/file_copy/test_copyfile.py b/lib/portage/tests/util/file_copy/test_copyfile.py
index e4979160f3..a3bb658ee3 100644
--- a/lib/portage/tests/util/file_copy/test_copyfile.py
+++ b/lib/portage/tests/util/file_copy/test_copyfile.py
@@ -2,6 +2,9 @@
 # Distributed under the terms of the GNU General Public License v2
 
+import contextlib
+import fcntl
 import shutil
 import tempfile
+from unittest.mock import patch
 
 import pytest
@@ -10,5 +13,5 @@ from portage import os
 from portage.tests import TestCase
 from portage.checksum import perform_md5
-from portage.util.file_copy import copyfile
+from portage.util.file_copy import copyfile, FICLONE
 
 
@@ -33,4 +36,5 @@ class CopyFileTestCase(TestCase):
 class CopyFileSparseTestCase(TestCase):
     def testCopyFileSparse(self):
+        stack = contextlib.ExitStack()
         tempdir = tempfile.mkdtemp()
         try:
@@ -50,5 +54,13 @@ class CopyFileSparseTestCase(TestCase):
                 # the end of the file.
                 f.truncate(2**18)
+            orig_ioctl = fcntl.ioctl
 
+            def mock_ioctl(fd, request, arg=0, mutate_flag=True):
+                result = orig_ioctl(fd, request, arg, mutate_flag)
+                if request == FICLONE:
+                    mock_ioctl.ficlone_success = True
+                return result
+
+            stack.enter_context(patch("fcntl.ioctl", new=mock_ioctl))
             copyfile(src_path, dest_path)
 
@@ -64,5 +76,8 @@ class CopyFileSparseTestCase(TestCase):
             # This is expected to fail when sparse copy is not implemented.
             if src_stat.st_blocks != dest_stat.st_blocks:
+                if getattr(mock_ioctl, "ficlone_success", None):
+                    pytest.fail("sparse copy failed with FICLONE")
                 pytest.xfail(reason="sparse copy is not implemented")
         finally:
             shutil.rmtree(tempdir)
+            stack.close()
```

Reply:
Yeah, a _fastcopy mock to detect OSError sounds pretty reasonable.

Reply:
Pushed a _fastcopy mock.